File: cudaq-benchmarks-main/benchmark.py
#!/usr/bin/env python3
import time
import argparse
import sys
import csv
import os

import cudaq
import numpy as np

import nvtx

# import the circuits from circuits.py
from circuits import QAOA, QuantumVolume, QFT, CounterfeitCoin, GHZ

CIRCUIT_CLASSES = [GHZ, QAOA, QuantumVolume, QFT, CounterfeitCoin]

def eprint(*args, **kwargs):
    """
    Print to stderr.
    """
    print(*args, **kwargs, file=sys.stderr)


def print_result(result, num_qubits):
    """
    Print a histogram of the result
    """
    if num_qubits > 10:
        eprint('WARNING: cannot print results, num_qubits > 10')
        return

    d = {i: 0 for i in range(2**num_qubits)}
    d.update({int(k, 2): v for k, v in result.items()})

    print('\n'.join(f'{k} {v}' for k, v in d.items()))

def summarize_times(ts):
    return {
        'mean': float(np.mean(ts)),
        'median': float(np.median(ts)),
        'min': float(np.min(ts)),
        'max': float(np.max(ts)),
        'stddev': float(np.std(ts)),
    }

def write_csv_row(path, row):
    fieldnames = [
        'implementation', 'circuit', 'backend', 'qubits', 'shots',
        'repetitions', 'warmups', 'seed',
        'sample_mean_seconds', 'sample_median_seconds', 'sample_min_seconds',
        'sample_max_seconds', 'sample_stddev_seconds',
        'total_mean_seconds', 'total_median_seconds', 'total_min_seconds',
        'total_max_seconds', 'total_stddev_seconds',
        'program_total_seconds',
    ]

    directory = os.path.dirname(path)
    if directory:
        os.makedirs(directory, exist_ok=True)

    exists = os.path.exists(path)
    with open(path, 'a', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        if not exists:
            writer.writeheader()
        writer.writerow(row)

def parse_qft_input_state(value):
    if value is None:
        return None
    if any(c not in '01' for c in value):
        raise ValueError('--qft-input-state must be a bit string, e.g. 01011')
    return [int(c) for c in value]

def parse_qaoa_weights(value):
    if value is None:
        return None
    try:
        weights = [int(item) for item in value.split(',') if item != '']
    except ValueError as exc:
        raise ValueError('--qaoa-weights must be a comma-separated list of -1/1') from exc
    if any(weight not in (-1, 1) for weight in weights):
        raise ValueError('--qaoa-weights may only contain -1 and 1')
    return weights

def run_one_experiment(num_qubits, num_shots, circuit_class, print_histo=False):
    program_start = time.perf_counter()
  
    if circuit_class is QFT and args.qft_input_state is not None:
        if len(args.qft_input_state) != num_qubits:
            raise ValueError('--qft-input-state length must equal --num-qubits')
        circuit = circuit_class(num_qubits, input_state=args.qft_input_state)
    elif circuit_class is QAOA and args.qaoa_weights is not None:
        expected = num_qubits * (num_qubits - 1) // 2
        if len(args.qaoa_weights) != expected:
            raise ValueError(f'--qaoa-weights length must be {expected}')
        circuit = circuit_class(num_qubits, rs=args.qaoa_weights)
    else:
        circuit = circuit_class(num_qubits)
    # --- instantiate circuits ---
    # circuit = QuantumVolume(num_qubits)
    # circuit = QAOA(num_qubits, np.pi/3, np.pi/6)
    # circuit = CounterfeitCoin(num_qubits)
    # circuit = QFT(num_qubits)
    #circuit = GHZ(num_qubits)

    # --- get kernel ---
    k, params = circuit.kernel, circuit.kernel_params

    sample_ts = []
    total_ts = []
    # --- main experiment loop ---
    for i in range(args.warmup + args.iter):
        r = nvtx.start_range(f'it{i}')
        total_t0 = time.perf_counter()
        sample_t0 = time.perf_counter()
        result = cudaq.sample(k, *params, 
                              shots_count=num_shots)
        sample_t1 = time.perf_counter()

        if print_histo:
            print(f'Most likely outcome: {result.most_probable()}')
            print_result(result, num_qubits)

        total_t1 = time.perf_counter()
        nvtx.end_range(r)
        
        sample_t = sample_t1 - sample_t0
        total_t = total_t1 - total_t0
        if i >= args.warmup:
            sample_ts.append(sample_t)
            total_ts.append(total_t)
        eprint(sample_t) # for debug

    sample_stats = summarize_times(sample_ts)
    total_stats = summarize_times(total_ts)
    avg = sample_stats['mean']
    std = sample_stats['stddev']
    print(f'{num_qubits} {avg} {std}')

    if args.csv_path:
        write_csv_row(args.csv_path, {
            'implementation': 'python',
            'circuit': circuit_class.__name__,
            'backend': args.target or '',
            'qubits': num_qubits,
            'shots': num_shots,
            'repetitions': args.iter,
            'warmups': args.warmup,
            'seed': args.seed if args.seed is not None else '',
            'sample_mean_seconds': sample_stats['mean'],
            'sample_median_seconds': sample_stats['median'],
            'sample_min_seconds': sample_stats['min'],
            'sample_max_seconds': sample_stats['max'],
            'sample_stddev_seconds': sample_stats['stddev'],
            'total_mean_seconds': total_stats['mean'],
            'total_median_seconds': total_stats['median'],
            'total_min_seconds': total_stats['min'],
            'total_max_seconds': total_stats['max'],
            'total_stddev_seconds': total_stats['stddev'],
            'program_total_seconds': time.perf_counter() - program_start,
        })

if __name__ == '__main__':
    # we need to set it manually, since CUDA-Q disables it by default
    sys.tracebacklimit = 1000

    p = argparse.ArgumentParser()
    p.add_argument('-n', '--num-qubits',   
                   type=int, required=True)
    p.add_argument('-N', '--num-qubits-max',
                   type=int, required=False,
                   help='if set, launches several repeat with `num_qubits` from `num_qubits` to `num_qubits_max`')
    p.add_argument('-s', '--num-shots',   
                   type=int, default=1024)
    p.add_argument('-w', '--warmup',
                   type=int, default=1,
                   help='numer of warmup iterations (default: 1)')
    p.add_argument('-i', '--iter',
                   type=int, default=10,
                   help='number of iterations (default: 10)')
    p.add_argument('--repetitions',
                   type=int, dest='iter',
                   help=argparse.SUPPRESS)
    # NOTE: `target` is parsed automatically by CUDA-Q, we do not need to handle it.
    p.add_argument('--target',
                   type=str, required=False,
                   help='target for CUDA-Q execution')

    circuit_names = [str(c.__name__) for c in CIRCUIT_CLASSES]
    p.add_argument('--circuit',
                   type=str, default=circuit_names[0], choices=circuit_names,
                   help=f'Circuit name to benchmark ({circuit_names}), default: {circuit_names[0]}.')
    p.add_argument('--seed',
                   type=int, required=False,
                   help='seed for random number generation (both NumPy and CUDA-Q)')
    p.add_argument('--histo',
                   action='store_true',
                   help='outputs a histogram of measurements for all quantum states')
    p.add_argument('--csv', '--output-csv',
                   dest='csv_path',
                   help='optional CSV path for summary timing results')
    p.add_argument('--qft-input-state',
                   type=parse_qft_input_state,
                   help='optional QFT input bit string used for Python/C++ parity')
    p.add_argument('--qaoa-weights',
                   type=parse_qaoa_weights,
                   help='optional comma-separated QAOA weights, each -1 or 1')

    args = p.parse_args()
    num_qubits_min = args.num_qubits
    num_qubits_max = (args.num_qubits_max or num_qubits_min) + 1
    num_shots  = args.num_shots
    circuit_class = CIRCUIT_CLASSES[circuit_names.index(args.circuit)]

    if args.seed is not None:
        np.random.seed(args.seed)
        cudaq.set_random_seed(args.seed)

    if args.target is not None:
        cudaq.set_target(args.target)

    eprint(args)

    for num_qubits in range(num_qubits_min, num_qubits_max):
        run_one_experiment(num_qubits, num_shots, circuit_class, args.histo)
    

File: cudaq-benchmarks-main/cpp/benchmark.cpp
#include "kernels.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <cstdint>
#include <fstream>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using Clock = std::chrono::steady_clock;

struct Options {
  std::string circuit = "GHZ";
  int qubit_count = 0;
  std::size_t shots = 1024;
  int repetitions = 10;
  int warmups = 1;
  std::string backend = "compile-time";
  std::string csv_path;
  std::uint64_t seed = 0;
  bool has_seed = false;
  std::optional<std::vector<int>> qft_input_state;
  std::optional<std::vector<int>> qaoa_weights;
};

struct TimedRun {
  double sample_seconds = 0.0;
  double total_seconds = 0.0;
  std::size_t result_size = 0;
};

struct Stats {
  double mean = 0.0;
  double median = 0.0;
  double min = 0.0;
  double max = 0.0;
  double stddev = 0.0;
};

void print_help(const char *program) {
  std::cout
      << "Usage: " << program << " --circuit GHZ -n 20 -s 1024 -i 10 [options]\n\n"
      << "Options:\n"
      << "  --circuit NAME          Circuit: GHZ, QFT, QAOA (default: GHZ)\n"
      << "  -n, --num-qubits N      Number of qubits (required)\n"
      << "  -s, --num-shots N       Number of shots (default: 1024)\n"
      << "      --shots N           Alias for --num-shots\n"
      << "  -i, --iter N            Timed repetitions (default: 10)\n"
      << "      --repetitions N     Alias for --iter\n"
      << "  -w, --warmup N          Warmup repetitions (default: 1)\n"
      << "      --target NAME       Backend label for output. The CUDA-Q C++ target\n"
      << "                          is selected when compiling with nvq++ --target.\n"
      << "      --backend NAME      Alias for --target\n"
      << "      --csv PATH          Append summary CSV row to PATH\n"
      << "      --output-csv PATH   Alias for --csv\n"
      << "      --seed N            Seed C++ host-side parameter generation\n"
      << "      --qft-input-state BITS\n"
      << "                          Optional QFT input bit string for parity runs\n"
      << "      --qaoa-weights CSV  Optional comma-separated -1/1 QAOA weights\n"
      << "  -h, --help              Show this help\n";
}

int parse_int(const std::string &name, const std::string &value) {
  std::size_t consumed = 0;
  const int parsed = std::stoi(value, &consumed);
  if (consumed != value.size())
    throw std::invalid_argument("invalid integer for " + name + ": " + value);
  return parsed;
}

std::size_t parse_size(const std::string &name, const std::string &value) {
  std::size_t consumed = 0;
  const auto parsed = std::stoull(value, &consumed);
  if (consumed != value.size())
    throw std::invalid_argument("invalid size for " + name + ": " + value);
  return static_cast<std::size_t>(parsed);
}

std::uint64_t parse_u64(const std::string &name, const std::string &value) {
  std::size_t consumed = 0;
  const auto parsed = std::stoull(value, &consumed);
  if (consumed != value.size())
    throw std::invalid_argument("invalid integer for " + name + ": " + value);
  return static_cast<std::uint64_t>(parsed);
}

std::vector<int> parse_qft_input_state(const std::string &value) {
  std::vector<int> bits;
  bits.reserve(value.size());
  for (const auto c : value) {
    if (c != '0' && c != '1')
      throw std::invalid_argument("--qft-input-state must be a bit string");
    bits.push_back(c == '1' ? 1 : 0);
  }
  return bits;
}

std::vector<int> parse_qaoa_weights(const std::string &value) {
  std::vector<int> weights;
  std::stringstream stream(value);
  std::string item;

  while (std::getline(stream, item, ',')) {
    if (item.empty())
      continue;
    const int weight = parse_int("--qaoa-weights", item);
    if (weight != -1 && weight != 1)
      throw std::invalid_argument("--qaoa-weights may only contain -1 and 1");
    weights.push_back(weight);
  }

  return weights;
}

Options parse_args(int argc, char **argv) {
  Options options;

  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    auto need_value = [&](const std::string &name) -> std::string {
      if (i + 1 >= argc)
        throw std::invalid_argument(name + " requires a value");
      return argv[++i];
    };

    if (arg == "-h" || arg == "--help") {
      print_help(argv[0]);
      std::exit(0);
    } else if (arg == "--circuit") {
      options.circuit = cudaq_bench::canonical_circuit_name(need_value(arg));
    } else if (arg == "-n" || arg == "--num-qubits") {
      options.qubit_count = parse_int(arg, need_value(arg));
    } else if (arg == "-s" || arg == "--num-shots" || arg == "--shots") {
      options.shots = parse_size(arg, need_value(arg));
    } else if (arg == "-i" || arg == "--iter" || arg == "--repetitions") {
      options.repetitions = parse_int(arg, need_value(arg));
    } else if (arg == "-w" || arg == "--warmup") {
      options.warmups = parse_int(arg, need_value(arg));
    } else if (arg == "--target" || arg == "--backend") {
      options.backend = need_value(arg);
    } else if (arg == "--csv" || arg == "--output-csv") {
      options.csv_path = need_value(arg);
    } else if (arg == "--seed") {
      options.seed = parse_u64(arg, need_value(arg));
      options.has_seed = true;
    } else if (arg == "--qft-input-state") {
      options.qft_input_state = parse_qft_input_state(need_value(arg));
    } else if (arg == "--qaoa-weights") {
      options.qaoa_weights = parse_qaoa_weights(need_value(arg));
    } else {
      throw std::invalid_argument("unknown argument: " + arg);
    }
  }

  if (options.qubit_count <= 0)
    throw std::invalid_argument("--num-qubits must be a positive integer");
  if (options.repetitions <= 0)
    throw std::invalid_argument("--iter/--repetitions must be positive");
  if (options.warmups < 0)
    throw std::invalid_argument("--warmup must be non-negative");
  if (options.shots == 0)
    throw std::invalid_argument("--num-shots must be positive");
  if (options.qft_input_state &&
      static_cast<int>(options.qft_input_state->size()) != options.qubit_count)
    throw std::invalid_argument(
        "--qft-input-state length must equal --num-qubits");
  if (options.qaoa_weights) {
    const int expected = options.qubit_count * (options.qubit_count - 1) / 2;
    if (static_cast<int>(options.qaoa_weights->size()) != expected)
      throw std::invalid_argument("--qaoa-weights length must equal n*(n-1)/2");
  }

  options.circuit = cudaq_bench::canonical_circuit_name(options.circuit);
  return options;
}

double elapsed_seconds(Clock::time_point start, Clock::time_point end) {
  return std::chrono::duration<double>(end - start).count();
}

Stats summarize(std::vector<double> values) {
  if (values.empty())
    return {};

  Stats stats;
  stats.min = *std::min_element(values.begin(), values.end());
  stats.max = *std::max_element(values.begin(), values.end());
  stats.mean = std::accumulate(values.begin(), values.end(), 0.0) / values.size();

  std::sort(values.begin(), values.end());
  const auto mid = values.size() / 2;
  stats.median = values.size() % 2 == 0 ? (values[mid - 1] + values[mid]) / 2.0
                                        : values[mid];

  double variance = 0.0;
  for (const auto value : values) {
    const double delta = value - stats.mean;
    variance += delta * delta;
  }
  stats.stddev = std::sqrt(variance / values.size());
  return stats;
}

TimedRun run_sample(const Options &options, const std::vector<int> &qft_input,
                    const std::vector<int> &qaoa_weights) {
  TimedRun timed;
  const auto total_start = Clock::now();
  const auto sample_start = Clock::now();

  if (options.circuit == "GHZ") {
    auto result = cudaq::sample(options.shots, cudaq_bench::GHZKernel{},
                                options.qubit_count);
    timed.result_size = result.size();
  } else if (options.circuit == "QFT") {
    auto result =
        cudaq::sample(options.shots, cudaq_bench::QFTKernel{}, qft_input);
    timed.result_size = result.size();
  } else if (options.circuit == "QAOA") {
    constexpr double gamma = 1.04719755119659774615;
    constexpr double beta = 0.52359877559829887308;
    auto result = cudaq::sample(options.shots, cudaq_bench::QAOAKernel{},
                                options.qubit_count, qaoa_weights, gamma, beta);
    timed.result_size = result.size();
  } else {
    throw std::invalid_argument("unsupported circuit: " + options.circuit);
  }

  const auto sample_end = Clock::now();
  const auto total_end = Clock::now();
  timed.sample_seconds = elapsed_seconds(sample_start, sample_end);
  timed.total_seconds = elapsed_seconds(total_start, total_end);
  return timed;
}

void write_csv(const Options &options, const Stats &sample_stats,
               const Stats &total_stats, double program_total_seconds) {
  if (options.csv_path.empty())
    return;

  const std::filesystem::path csv_path(options.csv_path);
  if (csv_path.has_parent_path())
    std::filesystem::create_directories(csv_path.parent_path());

  const bool exists = static_cast<bool>(std::ifstream(options.csv_path));
  std::ofstream csv(options.csv_path, std::ios::app);
  if (!csv)
    throw std::runtime_error("failed to open CSV path: " + options.csv_path);

  if (!exists) {
    csv << "implementation,circuit,backend,qubits,shots,repetitions,warmups,seed,"
        << "sample_mean_seconds,sample_median_seconds,sample_min_seconds,"
        << "sample_max_seconds,sample_stddev_seconds,total_mean_seconds,"
        << "total_median_seconds,total_min_seconds,total_max_seconds,"
        << "total_stddev_seconds,program_total_seconds\n";
  }

  csv << std::setprecision(17) << "cpp," << options.circuit << ','
      << options.backend << ',' << options.qubit_count << ',' << options.shots
      << ',' << options.repetitions << ',' << options.warmups << ','
      << (options.has_seed ? std::to_string(options.seed) : "") << ','
      << sample_stats.mean << ',' << sample_stats.median << ','
      << sample_stats.min << ',' << sample_stats.max << ','
      << sample_stats.stddev << ',' << total_stats.mean << ','
      << total_stats.median << ',' << total_stats.min << ',' << total_stats.max
      << ',' << total_stats.stddev << ',' << program_total_seconds << '\n';
}

} // namespace

int main(int argc, char **argv) {
  try {
    const auto program_start = Clock::now();
    const auto options = parse_args(argc, argv);

    const auto qft_input =
        options.qft_input_state
            ? *options.qft_input_state
            : cudaq_bench::make_qft_input_state(options.qubit_count,
                                                options.seed, options.has_seed);
    const auto qaoa_weights =
        options.qaoa_weights
            ? *options.qaoa_weights
            : cudaq_bench::make_qaoa_weights(options.qubit_count, options.seed,
                                             options.has_seed);

    std::vector<double> sample_times;
    std::vector<double> total_times;
    sample_times.reserve(options.repetitions);
    total_times.reserve(options.repetitions);

    std::size_t last_result_size = 0;
    for (int i = 0; i < options.warmups + options.repetitions; ++i) {
      const auto timed = run_sample(options, qft_input, qaoa_weights);
      last_result_size = timed.result_size;

      if (i >= options.warmups) {
        sample_times.push_back(timed.sample_seconds);
        total_times.push_back(timed.total_seconds);
      }

      std::cerr << timed.sample_seconds << '\n';
    }

    const auto sample_stats = summarize(sample_times);
    const auto total_stats = summarize(total_times);
    const auto program_total_seconds =
        elapsed_seconds(program_start, Clock::now());

    std::cout << std::setprecision(10) << options.qubit_count << ' '
              << sample_stats.mean << ' ' << sample_stats.stddev << '\n';
    std::cerr << "result_size=" << last_result_size << '\n';

    write_csv(options, sample_stats, total_stats, program_total_seconds);
  } catch (const std::exception &error) {
    std::cerr << "error: " << error.what() << '\n';
    return 1;
  }

  return 0;
}
